File: java/src/main/java/com/aerospike/helper/query/KeyRecordIterator.java
package com.aerospike.helper.query;

import java.io.Closeable;
import java.io.IOException;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import org.apache.log4j.Logger;

import com.aerospike.client.Key;
import com.aerospike.client.Record;
import com.aerospike.client.query.KeyRecord;
import com.aerospike.client.query.RecordSet;
import com.aerospike.client.query.ResultSet;

public class KeyRecordIterator implements Iterator<KeyRecord>, Closeable {
	private static Logger log = Logger.getLogger(KeyRecordIterator.class);
	private RecordSet recordSet;
	private ResultSet resultSet;
	private Iterator<KeyRecord> recordSetIterator;
	private Iterator<Object> resultSetIterator;
	private String namespace;
	
	public KeyRecordIterator(String namespace) {
		super();
		this.namespace = namespace;
	}

	public KeyRecordIterator(String namespace, RecordSet recordSet) {
		this(namespace);
		this.recordSet = recordSet;
		this.recordSetIterator = recordSet.iterator();
	}


	public KeyRecordIterator(String namespace, ResultSet resultSet) {
		this(namespace);
		this.resultSet = resultSet;
		this.resultSetIterator = resultSet.iterator();
		
	}

	@Override
	public void close() throws IOException {
		if (recordSet != null)
			recordSet.close();
		if (resultSet != null)
			resultSet.close();
	}

	@Override
	public boolean hasNext() {
		if (this.recordSetIterator != null)
			return this.recordSetIterator.hasNext();
		else
			return this.resultSetIterator.hasNext();
	}

	@Override
	public KeyRecord next() {
		KeyRecord keyRecord = null;

		if (this.recordSetIterator != null) {
			keyRecord = this.recordSetIterator.next();
		} else {
			Map<String, Object> map = (Map) this.resultSetIterator.next();
			Map<String,Object> meta = (Map<String, Object>) map.get("meta_data");
			map.remove("meta_data");
			Map<String,Object> binMap = new HashMap<String, Object>(map);
			if (log.isDebugEnabled()){
				for (Map.Entry<String, Object> entry : map.entrySet())
				{
					log.debug(entry.getKey() + " = " + entry.getValue());
				}
			}
			Long generation =  (Long) meta.get("generation");
			Long ttl =  (Long) meta.get("ttl");
			Record record = new Record(binMap, generation.intValue(), ttl.intValue());
			Key key = new Key(namespace, (String) meta.get("set_name"), (byte[]) meta.get("digest")); 
			keyRecord = new KeyRecord(key , record);
		}
		return keyRecord;
	}

	@Override
	public void remove() {
		
	}

}

File: csharp/AerospikeHelper/AerospikeHelper/Query/KeyRecordIterator.cs
using System;
using log4net;
using Aerospike.Client;

using System.Collections;
using System.Collections.Generic;
using Aerospike.Helper.Model;
using System.Runtime.CompilerServices;

namespace Aerospike.Helper.Query
{
	/// <summary>
	/// Iterator for traversing a collection of KeyRecords
	/// </summary>
	public class KeyRecordEnumerator : IEnumerator<KeyRecord>
	{
		private const String META_DATA = "meta_data";
		private const String SET_NAME = "set_name";
		private const String DIGEST = "digest";
		private const String EXPIRY = "expiry";
		private const String GENERATION = "generation";
		private static readonly ILog log = LogManager.GetLogger (typeof(KeyRecordEnumerator));
		private RecordSet recordSet;
		private ResultSet resultSet;
		private String ns;
		private KeyRecord singleRecord;
		private KeyRecord currentRecord;

		public KeyRecordEnumerator (String ns)
		{
			this.ns = ns;
		}

		public KeyRecordEnumerator (String ns, KeyRecord singleRecord) : this (ns)
		{

			this.singleRecord = singleRecord;

		}

		public KeyRecordEnumerator (String ns, RecordSet recordSet) : this (ns)
		{
			
			this.recordSet = recordSet;

		}

		public KeyRecordEnumerator (String ns, ResultSet resultSet) : this (ns)
		{
			this.resultSet = resultSet;

		}

		#region IEnumerator

		public void Dispose ()
		{
			Close ();
		}

		public bool MoveNext ()
		{
			bool hasNext = false;
			if (this.recordSet != null && this.recordSet.Next ()) {
				currentRecord = new KeyRecord (this.recordSet.Key, this.recordSet.Record);
				hasNext = true;

			} else if (this.resultSet != null && this.resultSet.Next ()) {
				Dictionary<String, Object> map = (Dictionary<String, Object>)this.resultSet.Object;
				Dictionary<String,Object> meta = (Dictionary<String, Object>)map [META_DATA];
				map.Remove (META_DATA);
				Dictionary<String,Object> binMap = new Dictionary<String, Object> (map);
				int generation = (int)meta [GENERATION];
				int ttl = (int)meta [EXPIRY];
				Record record = new Record (binMap, generation, ttl);
				Key key = new Key (ns, (byte[])meta [DIGEST], (String)meta [SET_NAME], null);
				currentRecord = new KeyRecord (key, record);
				hasNext = true;

			} else if (singleRecord != null) {
				currentRecord = singleRecord;
				singleRecord = null;
				hasNext = true;
			}

			return hasNext;		
		}

		public void Reset ()
		{
		}

		public KeyRecord Current {
			get { return currentRecord; }
		}

		object IEnumerator.Current {
			get { return Current; }
		}

		#endregion

		[MethodImpl (MethodImplOptions.Synchronized)]
		public void Close ()
		{
			if (recordSet != null)
				recordSet.Close ();
			if (resultSet != null)
				resultSet.Close ();
			singleRecord = null;
		}

		public override String ToString ()
		{
			return this.ns;
		}

	}
}

